File: 프로그래머스/lv3/49191. 순위/순위.py
def solution(n, results):
    answer = 0

    board = [[0] * (n + 1) for _ in range(n + 1)]

    for a, b in results:
        board[a][b]=1
        board[b][a]=-1


    for k in range(1, n + 1):
        for s in range(1, n + 1):
            for e in range(1, n + 1):
                if board[s][k] == board[k][e] == 1:
                    board[s][e] = 1 # 이김
                    board[e][s] = board[k][s] = board[e][k] = -1 # 짐
    
    for row in board:
        if row[1:].count(0) == 1: # 1 승리 /0 모름 / -1짐
            answer += 1
            
    return answer

File: 프로그래머스/3/49191. 순위/순위.java
class Solution {
    public int solution(int n, int[][] results) {
        int answer = 0;
        
        
        boolean[][] graph = new boolean[n][n];
        for (int[] edge: results){
            int u = edge[0]-1;
            int v = edge[1]-1;
            graph[u][v] = true;
        }
        
        
        for(int u=0; u<n; u++){
            int wins = countForward(u, graph, new boolean[n]) -1;
            int loses = countBackward(u, graph, new boolean[n]) -1;
            
            if (wins+loses+1==n){
                answer++;
            }

        }
        
        return answer;
    }
    
    private int countForward(int u, boolean[][] graph, boolean[] visited){ // 이긴 수
        int cnt = 1;
        
        for(int v=0; v<graph[u].length; v++){
            if (!graph[u][v] || visited[v]) continue;
            visited[v] = true;
            cnt+= countForward(v, graph, visited);
        }
        return cnt;
    }
    
        private int countBackward(int u, boolean[][] graph, boolean[] visited){ // 진 수
        int cnt = 1;
        
        for(int v=0; v<graph.length; v++){
            if (!graph[v][u] || visited[v]) continue;
            visited[v] = true;
            cnt+= countBackward(v, graph, visited);
        }
        return cnt;
    }
    
    
    
}